### PR TITLE
Add a new rendering engine

### PR DESCRIPTION
**Pull request by ihhub**
The work is in progress:
- add a proper selection of bit depth

### CODE REVIEW

**Comment by idshibanov** [comment]
Looks great to me so far. Obviously there's a lot of work still left to do, but I think you've got the tools to do it. We'll also need primitives drawing and image manipulation functions, for fade, blur and special effects.

File: src/fheroes2/gui/ui_button.h [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+    private:
+        int32_t _offsetX;
+        int32_t _offsetY;
+
+        int _icnId;
+        uint32_t _releasedIndex;
+        uint32_t _pressedIndex;
+
+        int _state;
```
Comment (Moiman) [inline src/fheroes2/gui/ui_button.h:56]: Why not just use two booleans (pressed and enabled) for state?

Comment (ihhub) [reply src/fheroes2/gui/ui_button.h:56]: Hi @Moiman , performance wise there is no difference between 2 methods. I didn't want to create many members of the class.

Comment (Moiman) [reply src/fheroes2/gui/ui_button.h:56]: I think booleans would make code way more readable than bitwise operations.

Comment (ihhub) [reply src/fheroes2/gui/ui_button.h]: Changed :)

**Comment by ihhub** [comment]
Hi @Sajjon , try to render Air Elemental and you will immediately see the difference.

**Comment by Sajjon** [comment]
@ihhub ah will do! But anyway, where does the transformTable come from? From Heroes2 AGG File like KB_PAL? Or did you calculate it yourself somehow?

**Pull request merged**